### PR TITLE
Support multiple providers for the same service

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: tests/services_multi_test.toit [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// ------------------------------------------------------------------
+
+class PingServiceClient extends services.ServiceClient implements PingService:
+  constructor --open/bool=true:
+    super --open=open
+
+  open -> PingServiceClient?:
+    return (open_ PingService.UUID PingService.MAJOR PingService.MINOR) and this
+
+  open [--filter] -> PingServiceClient?:
```
Comment (kasperl) [inline tests/services_multi_test.toit:70]: Still haven't figured out the best way to write the clients. I dislike having to write too much boilerplate here.

File: lib/system/services.toit [hunk 2]
```diff
@@ -50,25 +49,52 @@ abstract class ServiceClient:
...
-        pid = _client_.discover uuid false
-      if not pid: return null
+      discovered = _client_.discover uuid false
+    if not discovered: return null
+
+    candidate_index := null
+    candidate_priority := null
+    for i := 0; i < discovered.size; i += 4:
+      tags := discovered[i + 3]
+      if filter.call tags:
```
Comment (kasperl) [inline lib/system/services.toit:69]: I'll probably change this so the filter receives the name, version, and tags independently -- possibly wrapped in a nice object.

File: lib/core/print.toit [hunk 3]
```diff
@@ -83,8 +83,8 @@ write_on_stderr_ message/string add_newline/bool -> none:
 /**
 Print service used by $print.
 */
-service_/PrintService ::= (PrintServiceClient --no-open).open or
-    StandardPrintService_
+service_/PrintService ::=
+    ((PrintServiceClient).open --if_absent=: StandardPrintService_) as PrintService
```
Comment (kasperl) [inline lib/core/print.toit:87]: Still not quite happy with this.

File: lib/system/api/print.toit [hunk 4]
```diff
@@ -2,22 +2,23 @@
 // Use of this source code is governed by an MIT-style license that can be
 // found in the lib/LICENSE file.
 
-import system.services show ServiceClient
+import system.services show ServiceSelector ServiceClient
 
 interface PrintService:
   static UUID  /string ::= "0b7e3aa1-9fc9-4632-bb09-4605cd11897e"
```
Comment (kasperl) [inline lib/system/api/print.toit:8]: These can go away when there are no users left of them.

**Review by floitsch — APPROVED** [review]
LGTM.

File: tests/services_multi_test.toit [hunk 5]
```diff
@@ -0,0 +1,259 @@
+// Copyright (C) 2023 Toitware ApS.
+// Use of this source code is governed by a Zero-Clause BSD license that can
+// be found in the tests/LICENSE file.
+
+import system.services
+import expect
```
[the remark this answers is not in the document]
Comment (kasperl) [reply tests/services_multi_test.toit]: Done.

File: tests/services_multi_test.toit [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    tests++
+
+  selector.allow --name="ping/A" --major=11 --minor=22
+  selector.allow --name="ping/A" --major=1 --minor=22
+  selector.allow --name="ping/A" --major=11 --minor=2
+  with_client selector: | client/PingServiceClient |
+    expect.expect_equals "ping/B" client.name
+    tests++
+
+  expect.expect_equals 16 tests
```
[the remark this answers is not in the document]
Comment (kasperl) [reply tests/services_multi_test.toit]: Done.

File: tests/services_multi_test.toit [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+  with_client (PingService.SELECTOR.restrict.deny --tag="A"): | client/PingServiceClient |
+    expect.expect_equals "ping/B" client.name
+    tests++
+
+  with_client (PingService.SELECTOR.restrict.deny --name="ping/B" --major=3): | client/PingServiceClient |
+    expect.expect_equals "ping/A" client.name
+    tests++
+
+
+
```
[the remark this answers is not in the document]
Comment (kasperl) [reply tests/services_multi_test.toit]: Done.

File: lib/system/api/print.toit (hunk 4 above)
Comment (kasperl) [reply lib/system/api/print.toit]: Removed them already.

File: lib/core/print.toit (hunk 3 above)
Comment (kasperl) [reply lib/core/print.toit]: Found something (slightly) better.

File: tests/services_multi_test.toit (hunk 1 above)
Comment (kasperl) [reply tests/services_multi_test.toit]: Looks better now.

File: lib/system/services.toit (hunk 2 above)
Comment (kasperl) [reply lib/system/services.toit]: Done.

**Pull request merged**